pkg/ottl/contexts/internal/ctxutil/slice.go

```diff
@@ -5,61 +5,258 @@ package ctxutil // import "github.com/open-telemetry/opentelemetry-collector-con
+		return errMissingSetKey
 	}
 
-	i, err := keys[0].Int(ctx, tCtx)
+	idx, err := getSliceIndexFromKeys(ctx, tCtx, s.Len(), keys)
 	if err != nil {
 		return err
 	}
-	if i == nil {
-		resInt, err := FetchValueFromExpression[K, int64](ctx, tCtx, keys[0])
-		if err != nil {
-			return fmt.Errorf("unable to resolve an integer index in slice: %w", err)
+
+	return SetIndexableValue[K](ctx, tCtx, s.At(idx), val, keys[1:])
+}
+
+// CommonTypedSlice is an interface for typed pdata slices, such as pcommon.StringSlice,
+// pcommon.Int64Slice, pcommon.Int32Slice, etc.
+type CommonTypedSlice[T any] interface {
+	At(int) T
+	Len() int
+	FromRaw(val []T)
+	SetAt(int, T)
+	AsRaw() []T
+}
+
+// GetCommonTypedSliceValue is like GetSliceValue, but for retrieving a value from a pdata
+// typed slice. [V] is the type of the slice elements. If no keys are provided, it returns
+// an error. This function does not support slice elements indexing.
+func GetCommonTypedSliceValue[K, V any](ctx context.Context, tCtx K, s CommonTypedSlice[V], keys []ottl.Key[K]) (V, error) {
+	if len(keys) == 0 {
+		return *new(V), errMissingGetKey
+	} else if len(keys) > 1 {
+		return *new(V), fmt.Errorf(typeNotIndexableError, s)
+	}
+
+	idx, err := getSliceIndexFromKeys(ctx, tCtx, s.Len(), keys)
+	if err != nil {
+		return *new(V), err
+	}
+
+	return s.At(idx), nil
+}
+
+// SetCommonTypedSliceValue sets the value of a pdata typed slice element. [V] is the type
+// of the slice elements. The any-wrapped value type must be [V], otherwise an error is
+// returned. This function does not support slice elements indexing.
+func SetCommonTypedSliceValue[K, V any](ctx context.Context, tCtx K, s CommonTypedSlice[V], keys []ottl.Key[K], val any) error {
+	if len(keys) == 0 {
+		return errMissingSetKey
+	} else if len(keys) > 1 {
+		return fmt.Errorf(typeNotIndexableError, s)
+	}
+
+	idx, err := getSliceIndexFromKeys(ctx, tCtx, s.Len(), keys)
+	if err != nil {
+		return err
+	}
+
+	typeVal, ok := val.(V)
+	if !ok {
+		return fmt.Errorf("invalid value type provided for a slice of %T: %T", *new(V), val)
+	}
+
+	s.SetAt(idx, typeVal)
+	return nil
+}
+
+// SetCommonTypedSliceValues sets the value of a pdata typed slice. It does handle all
+// different input types OTTL generate, such as []V, []any, or a pcommon.Slice.
+// If the value is a slice of [any] or pcommon.Slice, and it has an element that the type
+// is not [V], an error is returned.
+func SetCommonTypedSliceValues[V any](s CommonTypedSlice[V], val any) error {
+	switch typeVal := val.(type) {
+	case CommonTypedSlice[V]:
+		s.FromRaw(typeVal.AsRaw())
+	case []V:
+		s.FromRaw(typeVal)
+	case []any:
+		raw := make([]V, len(typeVal))
+		for i, v := range typeVal {
+			sv, ok := v.(V)
+			if !ok {
+				return fmt.Errorf("invalid value type provided for a slice of %T: %T", *new(V), v)
+			}
+			raw[i] = sv
 		}
-		i = resInt
+		s.FromRaw(raw)
+	case pcommon.Slice:
+		raw := make([]V, typeVal.Len())
+		for i := 0; i < typeVal.Len(); i++ {
```

Comment:
Evaluation of `Len()` is only required once at loop start.
```suggestion
		for i := range typeVal.Len() {
```